### PR TITLE
[Bug Fix] Zone Heading for Binds, Summons, Teleports, and Zoning.

### PR DESCRIPTION
**Pull request by KinglyKrab**
For as long as I can remember people have had issues with zoning in,facing the wrong way, and walking through a zone line.

With this we will be able to set zone's safe heading as well as preserve heading on summon (NPC or GM) and teleports between zones.

This affects several pre-existing quest methods and extends their parameters to allow for the addition of heading.

The following functions have had heading added.

Lua
- client:SetBindPoint()
- client:SetStartZone()

Perl
- $client->SetBindPoint()
- $client->SetStartZone()
- quest::rebind()

SetStartZone parameter list was fixed also.

This converts some pre-existing methods from glm::vec3() to glm::vec4() and has an overload where necessary to use a glm::vec3() method versus glm::vec4() method.

This shouldn't affect any pre-existing servers and will allow PEQ and others to document safe headings for zones properly.

### CODE REVIEW

File: zone/client_packet.cpp [hunk 1]
```diff
@@ -12725,13 +12761,13 @@ void Client::Handle_OP_SetStartCity(const EQApplicationPacket *app)
...
+			zone_id = atoi(row[1]);
 		else
-			zoneid = atoi(row[0]);
+			zone_id = atoi(row[0]);
 
-		char* name = nullptr;
-		content_db.GetZoneLongName(ZoneName(zoneid), &name);
-		Message(Chat::Yellow, "%d - %s", zoneid, name);
+		char* zone_long_name = nullptr;
+		content_db.GetZoneLongName(ZoneName(zone_id), &zone_long_name);
```
[the remark this answers is not in the document]
Comment (KinglyKrab) [reply zone/client_packet.cpp:12769]: Yeah, I was looking at that and wondered why we did that. I'll change it.

File: zone/client_packet.cpp [hunk 2]
```diff
@@ -12725,13 +12761,12 @@ void Client::Handle_OP_SetStartCity(const EQApplicationPacket *app)
...
+			zone_id = atoi(row[1]);
 		else
-			zoneid = atoi(row[0]);
+			zone_id = atoi(row[0]);
 
-		char* name = nullptr;
-		content_db.GetZoneLongName(ZoneName(zoneid), &name);
-		Message(Chat::Yellow, "%d - %s", zoneid, name);
+		std::string zone_long_name = zone_store.GetZoneLongName(zone_id);
+		Message(Chat::Yellow, "%d - %s", zone_id, zone_long_name);
```
Comment (hgtw) [inline zone/client_packet.cpp:12769]: Needs to be `zone_long_name.c_str()` in the log message here

File: zone/zoning.cpp [hunk 3]
```diff
@@ -189,41 +199,54 @@ void Client::Handle_OP_ZoneChange(const EQApplicationPacket *app) {
...
+	float target_x = 0, target_y = 0, target_z = 0, target_heading = 0;
 	switch(zone_mode) {
 	case EvacToSafeCoords:
 	case ZoneToSafeCoords:
-		LogDebug("Zoning [{}] to safe coords ([{}],[{}],[{}]) in [{}] ([{}])", GetName(), safe_x, safe_y, safe_z, target_zone_name, target_zone_id);
-		dest_x = safe_x;
-		dest_y = safe_y;
-		dest_z = safe_z;
+		LogDebug(
+			"Zoning [{}] to safe coords ([{}], [{}], [{}]) in [{}] ([{}])",
```
Comment (hgtw) [inline zone/zoning.cpp:210]: Missing a format specifier `[{}]` for the new `safe_heading` arg

File: zone/zoning.cpp [hunk 4]
```diff
@@ -982,26 +1039,28 @@ void Client::SendZoneFlagInfo(Client *to) const {
...
+			&min_level,
+			flag_name
+		)) {
 			strcpy(flag_name, "(ERROR GETTING NAME)");
 		}
 
-		to->Message(Chat::White, "Has Flag %s for zone %s (%d,%s)", flag_name, long_name, zoneid, short_name);
-		if(long_name != empty)
-			delete[] long_name;
+		to->Message(Chat::White, "Has Flag %s for zone %s (%d,%s)", flag_name, zone_long_name, zone_id, zone_short_name);
```
Comment (hgtw) [inline zone/zoning.cpp:1063]: Needs to be `zone_long_name.c_str()` in the log message

File: utils/sql/db_update_manifest.txt [hunk 5]
```diff
@@ -416,6 +416,7 @@
 9160|2021_02_14_npc_exp_mod.sql|SHOW COLUMNS from `npc_types` LIKE 'exp_mod'|empty|
 9161|2021_02_15_npc_spell_entries_unsigned.sql|SELECT * FROM db_version WHERE version >= 9161|empty|
 9162|2021_02_17_server_scheduled_events.sql|SELECT * FROM db_version WHERE version >= 9162|empty|
+9163|2021_04_17_zone_safe_heading_changes.sql|SHOW COLUMNS FROM `zone` LIKE 'safe_heading'|empty|
```
Comment (hgtw) [inline utils/sql/db_update_manifest.txt:419]: The sql file is empty

Comment (KinglyKrab) [reply utils/sql/db_update_manifest.txt:419]: Fixed.

File: zone/zoning.cpp (hunk 4 above)
Comment (KinglyKrab) [reply zone/zoning.cpp:1063]: Fixed.

File: zone/zoning.cpp (hunk 3 above)
Comment (KinglyKrab) [reply zone/zoning.cpp]: Fixed.

File: zone/client_packet.cpp (hunk 2 above)
Comment (KinglyKrab) [reply zone/client_packet.cpp]: Fixed.

**Review by hgtw — APPROVED** [review]

File: zone/perl_client.cpp [hunk 6]
```diff
@@ -771,47 +771,28 @@ XS(XS_Client_SetEXP) {
...
 	{
 		Client *THIS;
-		int   to_zone;
-		int   to_instance;
-		float new_x;
-		float new_y;
-		float new_z;
+		int   to_zone = -1;
+		int   to_instance = 0;
+		float new_x = 0.0f, new_y = 0.0, new_z = 0.0, new_heading = 0.0f;
```
Comment (mackal) [inline zone/perl_client.cpp:780]: more Fs!

Comment (KinglyKrab) [reply zone/perl_client.cpp]: Fixed.

**Review by mackal — APPROVED** [review]

**Review by joligario — CHANGES_REQUESTED** [review]
With NOT NULL, recommend a default value.

**Review by joligario — APPROVED** [review]

File: world/worlddb.cpp [hunk 7]
```diff
@@ -54,334 +54,383 @@ void WorldDatabase::GetCharSelectInfo(uint32 account_id, EQApplicationPacket **o
...
+					`player_class` = {}
+					AND
+					`player_deity` = {}
+					AND
+					`player_race` = {} {}
+				),
 				p_character_select_entry_struct->Class,
 				p_character_select_entry_struct->Deity,
 				p_character_select_entry_struct->Race,
 				ContentFilterCriteria::apply().c_str()
```
Comment (Akkadius) [inline world/worlddb.cpp:230]: `c_str()` no longer becomes needed here when using `fmt::format`

File: world/worlddb.cpp [hunk 8]
```diff
@@ -398,8 +447,27 @@ int WorldDatabase::MoveCharacterToBind(int CharID, uint8 bindnum)
 		heading = atof(row[5]);
 	}
 
-	query = StringFormat("UPDATE character_data SET zone_id = '%d', zone_instance = '%d', x = '%f', y = '%f', z = '%f', heading = '%f' WHERE id = %u",
-						 zone_id, instance_id, x, y, z, heading, CharID);
+	query = fmt::format(
```
Comment (Akkadius) [inline world/worlddb.cpp:450]: Using repositories would clean stuff like this up big time, not needed for this PR though

**Review by Akkadius — APPROVED** [review]
LGTM; lots of good cleanups too

**Pull request merged**